### ISSUE TITLE
How can I pass --registry-mirror option to library/docker:dind ?

### ISSUE THREAD

**Issue by ihoro** [docker-library/docker#38]
It’s all about particular CI case, when docker:dind is used and it would be great to pass it something like env variable in order to make dockerd use local caching registry.

What do you recommend?

**Comment by phigoro** [comment]
e.g. DOCKER_DRIVER=overlay passed as environment variable works.

**Comment by phigoro** [comment]
I forgot to mention that the only option I have is env variable, due to it’s about GitLab CI when I can only define image name and env variables, i.e. I cannot pass parameters to the entry point.

**Comment by tianon** [comment]
I definitely recommend finding or filing an issue with GitLab CI to let them know that changing `command` is definitely something that's necessary for many use cases.

To work around this, I'd recommend creating a new image `FROM` this one which adds a simple startup script to read the environment variables you need in order to set the expected parameters for `dockerd`, which then just invokes `exec dockerd-entrypoint.sh --flags --you --need`.  (By default, `DOCKER_DRIVER` shouldn't be working -- you might want to double check what you're actually running, because it's likely using our default value of `vfs`, which is not going to be terribly performant and is going to use an excessive amount of disk space).

**Comment by tianon** [comment]
Doh, regarding `DOCKER_DRIVER`, looks like that's a Docker-specific feature: https://github.com/docker/docker/blob/db657a558ecff8e64490ebef15d744af7ef19d50/daemon/daemon.go#L544-L547
(added way back in https://github.com/docker/docker/pull/2609)

**Comment by phigoro** [comment]
Thanks for comments. Yes, I tried custom image based on docker:dind, but then it leads to another problem of GitLab CI :) Both issues are related to “services” in GitLab CI and there are respective bugs open already. I simply wanted to hear your opinion whether you have a vision of some workaround until all these things get fixed on GitLab CI side.

Yeah, DOCKER_DRIVER is supported directly, but it’s strange that other options are not supported such way. I believe there are arguments why, but from quick overview it looks inconvenient, e.g. it would solve my case.

**Comment by mitar** [comment]
I would love this as well.

**Comment by yosifkit** [comment]
FYI, you can specify `command` on GitLab CI now: https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#available-settings-for-services so there is no need to use env vars to hack around their limitations.

**Comment by relax-admin** [comment]
Docker in docker image allowing to set up daemon configuration via environment variables

Available variables:

```
DOCKER_REGISTRY_MIRROR => --registry-mirror
DOCKER_INSECURE_REGISTRY => --insecure-registry
DOCKER_MAX_CONCURRENT_DOWNLOADS => --max-concurrent-downloads
```

for gitlab
```yml
  services:
    - name: $PRIVATE_REGISTRY/docker:19.03.1-dind
      command: ["--insecure-registry=registry.elandsystems.cn:5000","--registry-mirror=https://dockerhub.azk8s.cn" ]
      alias: docker
```

**Comment by Yannik** [comment]
> Docker in docker image allowing to set up daemon configuration via environment variables
> 
> Available variables:
> 
> ```
> DOCKER_REGISTRY_MIRROR => --registry-mirror
> DOCKER_INSECURE_REGISTRY => --insecure-registry
> DOCKER_MAX_CONCURRENT_DOWNLOADS => --max-concurrent-downloads
> ```

I couldn't find any references to these variables, and using them did not work for me.

Can you tell me where you got this from?

**Comment by tianon** [comment]
I think the point that was being made there is that each of those requested environment variables are trivially available via command-line arguments to the image instead.

**Comment by weakcamel** [comment]
@relax-admin 

> Docker in docker image allowing to set up daemon configuration via environment variables

I searched through the Dockerfile od "dind" image and found nothing (see https://github.com/docker-library/docker)

And experimentally, it didn't work either, e.g.

```
docker run --privileged -d -v $PWD/certs:/certs  --name dind -p 2376:2376 -e DOCKER_REGISTRY_MIRROR=https://dockerhub.my.domain docker:19.03.5-dind

export DOCKER_HOST=tcp://localhost:2376
export DOCKER_TLS_VERIFY=1
export DOCKER_CERT_PATH=$PWD/certs/client/


# and this did *not* go through the mirror:
docker pull zshusers/zsh:4.3.15
```

**Comment by strowi** [comment]
+1 to the env variable. Especially with the recently introduced rate-limit of dockerhub, there needs to be an easy and globally possible way to pass a registry-mirror.  
In our case developers would have to add the "--registry-mirror=" command to every single job using dind. Instead it would be very comfortable to  just add a variable to the gitlab-runner by the infra/ci-team like:

```
[[runners]]
  name = "ci-gitlab-runner-xyz"
  url = ""
  token = ""
  executor = "kubernetes"
  environment = ["DOCKER_HOST=tcp://localhost:2375", "DOCKER_TLS_CERTDIR=", "DOCKER_REGISTRY_MIRROR="mirror.xyz.net" ]
```
This way every job would use the mirror automatically...

**Comment by strowi** [comment]
Hi,

+1 to the env variable. Especially with the recently announced rate-limit on dockerhub. Using Gitlab, with the command-arg every team would need to add the "--registry-mirror"-parameget to every job which is a lot of work and error prone).

If this were a variable it could be passed to ALL Jobs running on the kubernetes-executor as easily as:

```
[[runners]]
  name = "ci-gitlab-runner-xxx"
  url = "https://gitlab.xxx.net"
  token = ""
  executor = "kubernetes"
  environment = ["DOCKER_HOST=tcp://localhost:2375", "DOCKER_TLS_CERTDIR=", DOCKER_REGISTRY_MIRROR="https://mirror.xxx.net"]
```

This way nobody has to do extra-work (which will be forgotten!) to use the registry mirror.

**Comment by weakcamel** [comment]
@tianon Is there any chance this could be re-considered? Seeing the Docker Hub rate limiting coming soon, an option to configure registry mirror  in CI would be invaluable!

**Comment by ihoro** [comment]
If I've got it correctly it's better to keep it open not to confuse a seeker with "Closed" as solved completely. It could be another issue open specifically for env var concern, but, probably, it's better to keep all thoughts and advices together. Let me know if I'm wrong.

**Comment by strowi** [comment]
@tianon  I can understand not supporting every single parameter through an env-variable. The alternative would be sth like https://github.com/docker-library/docker/issues/38 a general DOCKER_OPTS=".." where you can specify whatever you need (and it could be pre-filled with the current defaults). 

What do you mean by "part of docker itself"?

**Comment by tianon** [comment]
> What do you mean by "part of docker itself"?

I mean a PR to https://github.com/moby/moby :smile:

**Comment by strowi** [comment]
Ah yes, totally agree. BTW the gitlab/k8s/dind seems to have been solved using a configmap and will be released with 13.6 as far is i understand -> https://gitlab.com/gitlab-org/gitlab/-/merge_requests/46612
